### PR TITLE
[AArch64][SVE2] Generate SVE2 BSL instruction in LLVM for add/sub.

### PR DESCRIPTION
**Pull request by dtemirbulatov**
Allow to fold or/and-and to BSL instuction for scalable vectors.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -17942,11 +17942,13 @@ static SDValue tryCombineToBSL(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
       } else
         continue;
 
-      if (!ISD::isBuildVectorAllZeros(Sub.getOperand(0).getNode()))
+      if (!ISD::isBuildVectorAllZeros(Sub.getOperand(0).getNode()) &&
+          !ISD::isConstantSplatVectorAllZeros(Sub.getOperand(0).getNode()))
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17946]: I believe isBuildVectorAllZeros is just a sub-set of isConstantSplatVectorAllZeros (it just calls isConstantSplatVectorAllZeros with BuildVectorOnly=true). So hopefully only the second is needed.

File: llvm/test/CodeGen/AArch64/sve2-bitselect.ll [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+; Check that an expanded vbsl(vneg(pre_cond), left, right) lowers to a VBSL
+; during ISEL.
+;
+; Subtly different from a plain vector bit select: operand representing the
+; condition has been negated (-v, not to be confused with bitwise_not(v)).
+
+; Each vbsl_neg_cond_xxxx tests one of the 16 permutations of the operands.
+
+define <vscale x 4 x i32> @vbsl_neg_cond_0000(<vscale x 4 x i32> %pre_cond, <vscale x 4 x i32> %left, <vscale x 4 x i32> %right) #0 {
```
Comment (david-arm) [inline llvm/test/CodeGen/AArch64/sve2-bitselect.ll:14]: Given all the tests in this file are repeats of those in neon-bitselect.ll I wonder if it's necessary to test all permutations? The NEON tests cover all cases and the code is the same for both NEON and SVE so perhaps we only need `@vbsl_neg_cond_0000` as a test?

Also, it's probably worth adding at least one test for the other DAG combine - (or (and a b) (and (not a) c)) => (bsl a b c). I think there are some NEON examples of this in neon-bitwise-instructions.ll

**Comment by david-arm** [comment]
This is just a thought, but SVE2 also has NBSL which is essentially just NOT(BSL). Is there any value in adding new patterns for this too?

**Comment by dtemirbulatov** [comment]
> This is just a thought, but SVE2 also has NBSL which is essentially just NOT(BSL). Is there any value in adding new patterns for this too?

yes, I think I could do it in the follow-up patch.

File: llvm/test/CodeGen/AArch64/sve2-bitselect.ll (hunk 2 above)
Comment (dtemirbulatov) [reply llvm/test/CodeGen/AArch64/sve2-bitselect.ll]: Done.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 1 above)
Comment (dtemirbulatov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Done.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 3]
```diff
@@ -17942,11 +17942,13 @@ static SDValue tryCombineToBSL(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
       } else
         continue;
 
-      if (!ISD::isBuildVectorAllZeros(Sub.getOperand(0).getNode()))
+      if (!ISD::isConstantSplatVectorAllZeros(Sub.getOperand(0).getNode()) &&
+          !ISD::isConstantSplatVectorAllZeros(Sub.getOperand(0).getNode()))
```
Comment (david-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17946]: Isn't the second call identical to the first? I think you only need one.

Comment (dtemirbulatov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17946]: Done.

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Pull request merged**